raven/src/main/java/com/getsentry/raven/buffer/DiskBuffer.java

```diff
@@ -0,0 +1,215 @@
+package com.getsentry.raven.buffer;
+
+import com.getsentry.raven.event.Event;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.Arrays;
+import java.util.Iterator;
+
+/**
+ * Stores {@link Event} objects to a directory on the filesystem and allows
+ * them to be flushed to Sentry (and deleted) at a later time.
+ */
+public class DiskBuffer implements Buffer {
+
+    /**
+     * File suffix added to all serialized event files.
+     */
+    public static final String FILE_SUFFIX = ".raven-event";
+
+    private static final Logger logger = LoggerFactory.getLogger(DiskBuffer.class);
+
+    private int maxEvents;
+    private final File bufferDir;
+
+    /**
+     * Construct an DiskBuffer which stores errors in the specified directory on disk.
+     *
+     * Private, you should use {@link DiskBuffer#newDiskBuffer(File, int)} to create
+     * an instance.
+     *
+     * @param bufferDir File representing directory to store buffered Events in
+     * @param maxEvents The maximum number of events to store offline
+     */
+    protected DiskBuffer(File bufferDir, int maxEvents) {
+        super();
+
+        this.bufferDir = bufferDir;
+        this.maxEvents = maxEvents;
+
+        logger.debug(Integer.toString(getNumStoredEvents())
+            + " stored events found in dir: "
+            + bufferDir.getAbsolutePath());
+    }
+
+    /**
+     * Attempt to create a new {@link DiskBuffer} instance backed by the provided bufferDir. If the
+     * bufferDir cannot be created or written to, this returns a {@link NoopBuffer} instance.
+     *
+     * @param bufferDir File representing the directory to write buffered Events to
+     * @param maxEvents the maximum number of Events to store in the bufferDir
+     * @return DiskBuffer if possible, otherwise NoopBuffer
+     */
+    public static Buffer newDiskBuffer(File bufferDir, int maxEvents) {
+        String errMsg = "Could not create or write to " + bufferDir.toString() + ", using Noop buffer instance.";
+
+        try {
+            bufferDir.mkdirs();
+            if (bufferDir.isDirectory() && bufferDir.canWrite()) {
+                return new DiskBuffer(bufferDir, maxEvents);
+            } else {
+                logger.error(errMsg);
+            }
+        } catch (Exception e) {
+            logger.error(errMsg, e);
+        }
+
+        return new NoopBuffer();
```

Comment:
Yeah, this idea came down from Android. I think you're right about throwing, and we can take a different route in Android (maybe). I'll scrap this stuff.